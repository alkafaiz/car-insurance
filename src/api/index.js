export const api = {
  submit: (url, success = true) =>
    new Promise(resolve => {
      setTimeout(function() {
        const response =
          success === false
            ? { success: false, message: "Unable to process requests." }
            : { success: true };

        resolve(response);
      }, 2000);
    })
};
