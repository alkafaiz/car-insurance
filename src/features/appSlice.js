import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import { api } from "../api";

const initialState = {
  stepIndex: 0,
  data: {
    firstName: "",
    lastName: "",
    dob: "",
    email: "",
    plateNumber: "",
    hasClaim: "yes",
    licenseYear: "0",
    carMake: "",
    carModel: "",
    carManufacturerDate: ""
  },
  isLoading: false,
  success: false
};

export const submit = createAsyncThunk(
  "app/submit",
  async (payload, thunkApi) => {
    const { plateNumber } = thunkApi.getState().app.data;

    const mockFailed = plateNumber === "111";

    if (mockFailed) {
      const response = await api.submit("/submit", false);
      return thunkApi.rejectWithValue(response);
    } else {
      const response = await api.submit("/submit");
      return response;
    }
  }
);

export const appSlice = createSlice({
  name: "app",
  initialState,
  reducers: {
    setStepIndex: (state, action) => {
      state.stepIndex = action.payload;
    },
    setPersonalDetails: (state, action) => {
      state.data.firstName = action.payload.firstName;
      state.data.lastName = action.payload.lastName;
      state.data.dob = action.payload.dob;
      state.data.email = action.payload.email;
    },
    setCarDetails: (state, action) => {
      state.data.plateNumber = action.payload.plateNumber;
      state.data.hasClaim = action.payload.hasClaim;
      state.data.licenseYear = action.payload.licenseYear;
      state.data.carMake = action.payload.carMake;
      state.data.carModel = action.payload.carModel;
      state.data.carManufacturerDate = action.payload.carManufacturerDate;
    }
  },
  extraReducers: {
    [submit.pending]: state => {
      state.isLoading = true;
    },
    [submit.rejected]: state => {
      state.isLoading = false;
      state.success = false;
    },
    [submit.fulfilled]: state => {
      state.isLoading = false;
      state.success = true;
    }
  }
});

export const {
  setStepIndex,
  setPersonalDetails,
  setCarDetails
} = appSlice.actions;

export const appReducer = appSlice.reducer;
