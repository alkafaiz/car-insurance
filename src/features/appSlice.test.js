import store from "../app/store";
import * as slice from "./appSlice";
import { api } from "../api";

jest.mock("../api");

describe("Form actions", () => {
  const setPersonalDetailsPayload = {
    firstName: "John",
    lastName: "Doe",
    dob: new Date("2000-03-16T05:05:46.822Z").toISOString(),
    email: "johndoe@gmail.com"
  };

  const setCarDetailsPayload = {
    plateNumber: "1212",
    hasClaim: "yes",
    licenseYear: "2",
    carMake: "Honda",
    carModel: "City",
    carManufacturerDate: new Date("2003-03-16T05:05:46.822Z").toISOString()
  };

  it("should check setStepIndex action creator", () => {
    const payload = 1;
    const expectedAction = { type: slice.setStepIndex.type, payload };

    expect(slice.setStepIndex(payload)).toStrictEqual(expectedAction);
  });

  it("should check setPersonalDetails action creator", () => {
    const expectedAction = {
      type: slice.setPersonalDetails.type,
      payload: setPersonalDetailsPayload
    };

    expect(slice.setPersonalDetails(setPersonalDetailsPayload)).toStrictEqual(
      expectedAction
    );
  });

  it("should check setCarDetails action creator", () => {
    const expectedAction = {
      type: slice.setCarDetails.type,
      payload: setCarDetailsPayload
    };

    expect(slice.setCarDetails(setCarDetailsPayload)).toStrictEqual(
      expectedAction
    );
  });

  it("should setStepIndex action creator updates the store", () => {
    let state = store.getState().app;
    const index = 1;

    store.dispatch(slice.setStepIndex(index));
    state = store.getState().app;

    expect(state.stepIndex).toStrictEqual(index);
  });

  it("should setPersonalDetails action creator updates the store", () => {
    let state = store.getState().app;

    store.dispatch(slice.setPersonalDetails(setPersonalDetailsPayload));
    state = store.getState().app;

    const updatedPersonalDetail = {
      firstName: state.data.firstName,
      lastName: state.data.lastName,
      dob: state.data.dob,
      email: state.data.email
    };

    expect(updatedPersonalDetail).toStrictEqual(setPersonalDetailsPayload);
    expect(updatedPersonalDetail).toMatchSnapshot();
  });

  it("should setCarDetails action creator updates the store", () => {
    let state = store.getState().app;

    store.dispatch(slice.setCarDetails(setCarDetailsPayload));
    state = store.getState().app;

    const updatedCarDetail = {
      plateNumber: state.data.plateNumber,
      hasClaim: state.data.hasClaim,
      licenseYear: state.data.licenseYear,
      carMake: state.data.carMake,
      carModel: state.data.carModel,
      carManufacturerDate: state.data.carManufacturerDate
    };

    expect(updatedCarDetail).toStrictEqual(setCarDetailsPayload);
    expect(updatedCarDetail).toMatchSnapshot();
  });
});

describe("Submit thunk actions", () => {
  it("should check submit thunk action creator", async () => {
    store.dispatch(slice.setCarDetails({ plateNumber: "111" }));
    store.dispatch(slice.submit());

    expect(api.submit).toHaveBeenCalledTimes(1);
    expect(api.submit).toHaveBeenCalledWith("/submit", false);
  });
});
