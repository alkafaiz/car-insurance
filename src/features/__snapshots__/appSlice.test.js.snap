// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form actions should setCarDetails action creator updates the store 1`] = `
Object {
  "carMake": "Honda",
  "carManufacturerDate": "2003-03-16T05:05:46.822Z",
  "carModel": "City",
  "hasClaim": "yes",
  "licenseYear": "2",
  "plateNumber": "1212",
}
`;

exports[`Form actions should setPersonalDetails action creator updates the store 1`] = `
Object {
  "dob": "2000-03-16T05:05:46.822Z",
  "email": "johndoe@gmail.com",
  "firstName": "John",
  "lastName": "Doe",
}
`;
